src/service/fixed-layer.js

```diff
@@ -0,0 +1,417 @@
+    const stylesheets = document.styleSheets;
+    const fixedSelectors = [];
+    for (let i = 0; i < stylesheets.length; i++) {
+      const stylesheet = stylesheets[i];
+      if (stylesheet.disabled ||
+              !stylesheet.ownerNode ||
+              stylesheet.ownerNode.tagName != 'STYLE' ||
+              stylesheet.ownerNode.hasAttribute('amp-boilerplate') ||
+              stylesheet.ownerNode.hasAttribute('amp-runtime') ||
+              stylesheet.ownerNode.hasAttribute('amp-extension')) {
+        continue;
+      }
+      this.discoverFixedSelectors_(stylesheet.cssRules, fixedSelectors);
+    }
+
+    try {
+      const idMap = {};
+      fixedSelectors.forEach(selector => {
+        const elements = this.doc.querySelectorAll(selector);
+        for (let i = 0; i < elements.length; i++) {
+          if (i > 10) {
+            // We shouldn't have too many of `fixed` elements.
+            break;
+          }
+          this.setupFixedElement_(elements[i], selector, idMap);
+        }
+      });
+    } catch (e) {
+      // Fail quietly.
+      setTimeout(() => {throw e;});
+    }
+
+    // Sort in document order.
+    this.fixedElements_.sort(function(fe1, fe2) {
+      // 8 | 2 = 0x0A
+      // 2 - preceeding
+      // 8 - contains
+      if (fe1.element.compareDocumentPosition(fe2.element) & 0x0A != 0) {
+        return -1;
+      }
+      return 1;
+    });
+
+    this.update();
+  }
+
+  /**
+   * @param {number} paddingTop
+   */
+  updatePaddingTop(paddingTop) {
+    this.paddingTop_ = paddingTop;
+    this.update();
+  }
+
+  /**
+   * Performs fixed actions.
+   * 1. Updates `top` styling if necessary.
+   * 2. On iOS/Iframe moves elements between fixed layer and BODY depending on
+   * whether they are currently visible and fixed.
+   * @return {!Promise}
+   */
+  update() {
+    if (this.fixedElements_.length == 0) {
+      return Promise.resolve();
+    }
+
+    // Some of the elements may no longer be in DOM.
+    /** @type {!Array<number>} */
+    const toRemove = [];
+    this.fixedElements_.forEach((fe, i) => {
+      if (!this.doc.contains(fe.element)) {
+        toRemove.push(i);
+      }
+    });
+    if (toRemove.length > 0) {
+      toRemove.forEach(i => {
+        this.fixedElements_.splice(i, 1);
+      });
+    }
+
+    /** @type {!Object<string, !FixedElementStateDef>} */
+    const fixedMap = {};
+    let hasTransferables = false;
+    return this.vsync_.runPromise({
+      measure: () => {
+        this.fixedElements_.forEach(fe => {
+          const element = fe.element;
+          const styles = this.doc.defaultView.getComputedStyle(element, null);
+          const position = styles.getPropertyValue('position');
+          const top = styles.getPropertyValue('top');
+          const bottom = styles.getPropertyValue('bottom');
+          const isFixed = (
+              position == 'fixed' &&
+              element.offsetWidth > 0 &&
+              element.offsetHeight > 0);
```

Comment:
They don't. But there's no point of doing all the work of moving the element into a fixed layer if it's not even visible.